core/src/main/java/com/datastax/oss/driver/api/core/uuid/Uuids.java

```diff
@@ -202,13 +235,188 @@ private static long makeClockSeqAndNode() {
+   *     available on this platform.
+   */
+  @NonNull
+  public static UUID nameBased(@NonNull byte[] namespaceAndName) {
+    return nameBased(namespaceAndName, 3);
+  }
+
+  /**
+   * Creates a new name-based (version 3 or version 5) {@link UUID} from the given byte array
+   * containing the namespace UUID and the name parts concatenated together.
+   *
+   * <p>The byte array is expected to be at least 16 bytes long.
+   *
+   * @param namespaceAndName A byte array containing the concatenated namespace UUID and name;
+   *     cannot be null.
+   * @param version The version to use, must be either 3 or 5.
+   * @throws NullPointerException if <code>namespaceAndName</code> is null.
+   * @throws IllegalArgumentException if <code>namespaceAndName</code> is not at least 16 bytes
+   *     long.
+   * @throws IllegalArgumentException if <code>version</code> is not 3 nor 5.
+   * @throws IllegalStateException if the {@link MessageDigest} algorithm for the desired version is
+   *     not available on this platform.
+   */
+  @NonNull
+  public static UUID nameBased(@NonNull byte[] namespaceAndName, int version) {
+    Objects.requireNonNull(namespaceAndName, "namespaceAndName cannot be null");
+    if (namespaceAndName.length < 16) {
+      throw new IllegalArgumentException("namespaceAndName must be at least 16 bytes long");
+    }
+    MessageDigest md = newMessageDigest(version);
+    md.update(namespaceAndName);
+    return buildNamedUuid(md.digest(), version);
+  }
+
+  @NonNull
+  private static MessageDigest newMessageDigest(int version) {
+    if (version != 3 && version != 5) {
+      throw new IllegalArgumentException(
+          "Invalid name-based UUID version, expecting 3 or 5, got: " + version);
+    }
+    String algorithm = version == 3 ? "MD5" : "SHA-1";
+    try {
+      return MessageDigest.getInstance(algorithm);
+    } catch (NoSuchAlgorithmException e) {
+      throw new IllegalStateException(algorithm + " algorithm not available", e);
+    }
+  }
+
+  @NonNull
+  private static UUID buildNamedUuid(@NonNull byte[] data, int version) {
+    // clear and set version
+    data[6] &= (byte) 0x0f;
+    data[6] |= (byte) (version << 4);
+    // clear and set variant to IETF
+    data[8] &= (byte) 0x3f;
+    data[8] |= (byte) 0x80;
+    return fromBytes(data);
+  }
+
+  private static UUID fromBytes(byte[] data) {
+    // data longer than 16 bytes will be truncated as mandated by the specs
+    assert data.length >= 16;
+    long msb = 0;
+    for (int i = 0; i < 8; i++) {
+      msb = (msb << 8) | (data[i] & 0xff);
+    }
+    long lsb = 0;
+    for (int i = 8; i < 16; i++) {
+      lsb = (lsb << 8) | (data[i] & 0xff);
+    }
+    return new UUID(msb, lsb);
+  }
+
+  private static byte[] toBytes(UUID uuid) {
+    byte[] out = new byte[16];
+    long msb = uuid.getMostSignificantBits();
+    for (int i = 0; i < 8; i++) {
+      out[i] = (byte) ((msb >> ((7 - i) * 8)) & 0xff);
```

Comment:
Nit: I think the mask is unnecessary, when you cast to `byte` the higher-order bytes will be discarded.
```suggestion
      out[i] = (byte) (msb >> ((7 - i) * 8));
```